File: backend/AIService.php
<?php

class AIService {
    private $config;
    
    public function __construct($config) {
        $this->config = $config;
    }
    
    public function isEnabled() {
        return $this->config['ai']['enabled'] ?? false;
    }
    
    public function getAvailableProviders() {
        $providers = [];
        if (!$this->isEnabled() || !isset($this->config['ai']['providers'])) {
            return $providers;
        }
        
        foreach ($this->config['ai']['providers'] as $name => $provider) {
            if (($provider['enabled'] ?? false) && !empty($provider['api_key'] ?? '')) {
                $providers[$name] = [
                    'name' => $name,
                    'models' => $provider['models'] ?? []
                ];
            }
        }
        return $providers;
    }
    
    public function translate($text, $sourceLanguage, $targetLanguage, $context = [], $provider = null, $model = null) {
        if (!$this->isEnabled()) {
            throw new Exception('AI features are not enabled');
        }
        
        $provider = $provider ?? $this->config['ai']['default_provider'];
        $model = $model ?? $this->config['ai']['default_model'];
        
        $providerConfig = $this->config['ai']['providers'][$provider] ?? null;
        if (!$providerConfig || !$providerConfig['enabled']) {
            throw new Exception("Provider $provider is not available");
        }
        
        $prompt = $this->buildTranslationPrompt($text, $sourceLanguage, $targetLanguage, $context);
        
        return $this->makeAIRequest($provider, $model, $prompt, $providerConfig);
    }
    
    public function proofread($text, $language, $context = [], $provider = null, $model = null) {
        if (!$this->isEnabled()) {
            throw new Exception('AI features are not enabled');
        }
        
        $provider = $provider ?? $this->config['ai']['default_provider'];
        $model = $model ?? $this->config['ai']['default_model'];
        
        $providerConfig = $this->config['ai']['providers'][$provider] ?? null;
        if (!$providerConfig || !$providerConfig['enabled']) {
            throw new Exception("Provider $provider is not available");
        }
        
        $prompt = $this->buildProofreadingPrompt($text, $language, $context);
        
        return $this->makeAIRequest($provider, $model, $prompt, $providerConfig, true);
    }
    
    private function buildTranslationPrompt($text, $sourceLanguage, $targetLanguage, $context) {
        $contextStr = '';
        if (!empty($context)) {
            $contextStr = "\n\nContext (related strings):\n";
            foreach ($context as $item) {
                $contextStr .= "- Key: {$item['key']}, {$sourceLanguage}: \"{$item['source']}\", {$targetLanguage}: \"{$item['target']}\"\n";
            }
        }
        
        return "Translate the following text from {$sourceLanguage} to {$targetLanguage}. This is for a mobile/desktop application localization.

Source text: \"{$text}\"
Source language: {$sourceLanguage}
Target language: {$targetLanguage}
{$contextStr}

Instructions:
- Provide only the translated text, no explanations
- Consider the context of mobile/desktop application UI
- Use an informal, friendly tone appropriate for modern mobile/desktop applications.
- If the target language distinguishes between formal and informal second-person address (T/V distinction), always use the informal form.
- Never mix formal and informal address within the same translation.
- Use the corresponding informal verb conjugations and possessive forms.
- Be consistent throughout all translations.
- Keep placeholders and formatting intact if any
- Consider the context of mobile/desktop application UI
- Do not translate product names, trademarks, or proper nouns unless they have an established localized form in {$targetLanguage}
- Use the 'key' field to understand the context and purpose of each string
- Do not omit or add entries
- Ensure all output is valid, properly escaped JSON
- Do not modify the 'key' values
- If it's a technical term or brand name, consider if it should remain untranslated

Translation:";
    }
    
    private function buildProofreadingPrompt($text, $language, $context) {
        $contextStr = '';
        if (!empty($context)) {
            $contextStr = "\n\nContext (related strings in the same language):\n";
            foreach ($context as $item) {
                $contextStr .= "- Key: {$item['key']}, Text: \"{$item['text']}\"\n";
            }
        }
        
        return "Review the following localized text for a mobile/desktop application. Evaluate the quality and provide feedback.

Text to review: \"{$text}\"
Language: {$language}
{$contextStr}

Please respond with a JSON object containing:
{
  \"status\": \"good\" | \"wording\" | \"issue\",
  \"feedback\": \"explanation of any issues or suggestions\"
}

Status meanings:
- \"good\": The text is well-written and appropriate
- \"wording\": The text is understandable but could be improved with better wording
- \"issue\": There are serious problems (grammar, unclear meaning, inappropriate tone, etc.)

Consider:
- Grammar and spelling
- Clarity and naturalness
- Appropriateness for UI context
- Consistency with typical app terminology
- Cultural appropriateness

Response:";
    }
    
    private function makeAIRequest($provider, $model, $prompt, $providerConfig, $expectJson = false) {
        switch ($provider) {
            case 'openai':
            case 'openai_compatible':
                return $this->makeOpenAIRequest($model, $prompt, $providerConfig, $expectJson);
            case 'anthropic':
                return $this->makeAnthropicRequest($model, $prompt, $providerConfig, $expectJson);
            case 'zai':
                return $this->makeZAIRequest($model, $prompt, $providerConfig, $expectJson);
            default:
                throw new Exception("Unsupported provider: $provider");
        }
    }
    
    private function makeOpenAIRequest($model, $prompt, $config, $expectJson = false) {
        $headers = [
            'Content-Type: application/json',
            'Authorization: Bearer ' . $config['api_key']
        ];
        
        $data = [
            'model' => $model,
            'messages' => [
                [
                    'role' => 'user',
                    'content' => $prompt
                ]
            ],
            'max_tokens' => $expectJson ? 40000 : 500, // Increase tokens for batch operations
            'temperature' => 0.3
        ];
        
        if ($expectJson) {
            $data['response_format'] = ['type' => 'json_object'];
        }
        
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $config['base_url'] . '/chat/completions');
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, $expectJson ? 120 : 30); // Longer timeout for batch operations
        
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        
        if ($response === false || $httpCode !== 200) {
            throw new Exception('AI request failed: HTTP ' . $httpCode);
        }
        
        $result = json_decode($response, true);
        if (!$result || !isset($result['choices'][0]['message']['content'])) {
            throw new Exception('Invalid AI response format');
        }
        
        $content = trim($result['choices'][0]['message']['content']);
        
        if ($expectJson) {
            $decoded = json_decode($content, true);
            if (!$decoded) {
                throw new Exception('Invalid JSON response from AI');
            }
            return $decoded;
        }
        
        return $content;
    }
    
    private function makeAnthropicRequest($model, $prompt, $config, $expectJson = false) {
        $headers = [
            'Content-Type: application/json',
            'x-api-key: ' . $config['api_key'],
            'anthropic-version: 2023-06-01'
        ];
        
        if ($expectJson) {
            $prompt .= "\n\nIMPORTANT: Respond only with valid JSON, no other text.";
        }
        
        $data = [
            'model' => $model,
            'max_tokens' => $expectJson ? 40000 : 500, // Increase tokens for batch operations
            'temperature' => 0.3,
            'messages' => [
                [
                    'role' => 'user',
                    'content' => $prompt
                ]
            ]
        ];
        
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $config['base_url'] . '/v1/messages');
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, $expectJson ? 120 : 30); // Longer timeout for batch operations
        
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        
        if ($response === false || $httpCode !== 200) {
            throw new Exception('AI request failed: HTTP ' . $httpCode);
        }
        
        $result = json_decode($response, true);
        if (!$result || !isset($result['content'][0]['text'])) {
            throw new Exception('Invalid AI response format');
        }
        
        $content = trim($result['content'][0]['text']);
        
        if ($expectJson) {
            $decoded = json_decode($content, true);
            if (!$decoded) {
                throw new Exception('Invalid JSON response from AI');
            }
            return $decoded;
        }
        
        return $content;
    }

    private function makeZAIRequest($model, $prompt, $config, $expectJson = false) {
        $headers = [
            'Content-Type: application/json',
            'Authorization: Bearer ' . $config['api_key']
        ];

        $data = [
            'model' => $model,
            'messages' => [
                [
                    'role' => 'user',
                    'content' => $prompt
                ]
            ],
            'max_tokens' => $expectJson ? 40000 : 500,
            'temperature' => 0.3
        ];

        if ($expectJson) {
            $data['response_format'] = ['type' => 'json_object'];
        }

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $config['base_url'] . '/chat/completions');
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, $expectJson ? 120 : 30);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($response === false || $httpCode !== 200) {
            throw new Exception('Z.AI request failed: HTTP ' . $httpCode);
        }

        $result = json_decode($response, true);
        if (!$result || !isset($result['choices'][0]['message']['content'])) {
            throw new Exception('Invalid Z.AI response format');
        }

        $content = trim($result['choices'][0]['message']['content']);

        if ($expectJson) {
            $decoded = json_decode($content, true);
            if (!$decoded) {
                throw new Exception('Invalid JSON response from Z.AI');
            }
            return $decoded;
        }

        return $content;
    }

    public function buildContext($currentKey, $allStrings, $language, $maxItems = 5) {
        $context = [];
        $count = 0;
        
        foreach ($allStrings as $key => $stringData) {
            if ($count >= $maxItems || $key === $currentKey) continue;
            
            if (isset($stringData['localizations'][$language]['stringUnit']['value'])) {
                $context[] = [
                    'key' => $key,
                    'text' => $stringData['localizations'][$language]['stringUnit']['value']
                ];
                $count++;
            }
        }
        
        return $context;
    }
    
    public function buildTranslationContext($currentKey, $allStrings, $sourceLanguage, $targetLanguage, $maxItems = 5) {
        $context = [];
        $count = 0;
        
        foreach ($allStrings as $key => $stringData) {
            if ($count >= $maxItems || $key === $currentKey) continue;
            
            $sourceText = $stringData['localizations'][$sourceLanguage]['stringUnit']['value'] ?? '';
            $targetText = $stringData['localizations'][$targetLanguage]['stringUnit']['value'] ?? '';
            
            if (!empty($sourceText) && !empty($targetText)) {
                $context[] = [
                    'key' => $key,
                    'source' => $sourceText,
                    'target' => $targetText
                ];
                $count++;
            }
        }
        
        return $context;
    }
    
    public function batchTranslate($items, $sourceLanguage, $targetLanguage, $provider = null, $model = null) {
        if (!$this->isEnabled()) {
            throw new Exception('AI features are not enabled');
        }
        
        if (empty($items)) {
            return [];
        }
        
        $provider = $provider ?? $this->config['ai']['default_provider'];
        $model = $model ?? $this->config['ai']['default_model'];
        
        $providerConfig = $this->config['ai']['providers'][$provider] ?? null;
        if (!$providerConfig || !$providerConfig['enabled']) {
            throw new Exception("Provider $provider is not available");
        }
        
        $prompt = $this->buildBatchTranslationPrompt($items, $sourceLanguage, $targetLanguage);
        
        $result = $this->makeAIRequest($provider, $model, $prompt, $providerConfig, true);
        
        // Parse and validate the JSON response
        if (!isset($result['translations']) || !is_array($result['translations'])) {
            throw new Exception('Invalid batch translation response format');
        }
        
        return $result['translations'];
    }
    
    public function batchProofread($items, $language, $provider = null, $model = null) {
        if (!$this->isEnabled()) {
            throw new Exception('AI features are not enabled');
        }
        
        if (empty($items)) {
            return [];
        }
        
        $provider = $provider ?? $this->config['ai']['default_provider'];
        $model = $model ?? $this->config['ai']['default_model'];
        
        $providerConfig = $this->config['ai']['providers'][$provider] ?? null;
        if (!$providerConfig || !$providerConfig['enabled']) {
            throw new Exception("Provider $provider is not available");
        }
        
        $prompt = $this->buildBatchProofreadingPrompt($items, $language);
        
        $result = $this->makeAIRequest($provider, $model, $prompt, $providerConfig, true);
        
        // Parse and validate the JSON response
        if (!isset($result['reviews']) || !is_array($result['reviews'])) {
            throw new Exception('Invalid batch proofreading response format');
        }
        
        return $result['reviews'];
    }
    
    private function buildBatchTranslationPrompt($items, $sourceLanguage, $targetLanguage) {
        $itemsJson = json_encode($items, JSON_PRETTY_PRINT);
        
        return "Translate the following strings from {$sourceLanguage} to {$targetLanguage}. This is for a mobile/desktop application localization.

Input data structure:
{$itemsJson}

Instructions:
- Translate each 'text' field from {$sourceLanguage} to {$targetLanguage}
- Consider the context of mobile/desktop application UI
- Use an informal, friendly tone appropriate for modern mobile/desktop applications.
- If the target language distinguishes between formal and informal second-person address (T/V distinction), always use the informal form.
- Never mix formal and informal address within the same translation.
- Use the corresponding informal verb conjugations and possessive forms.
- Be consistent throughout all translations.
- Keep placeholders and formatting intact if any
- Consider the context of mobile/desktop application UI
- Do not translate product names, trademarks, or proper nouns unless they have an established localized form in {$targetLanguage}
- Use the 'key' field to understand the context and purpose of each string
- Do not omit or add entries
- Ensure all output is valid, properly escaped JSON
- Do not modify the 'key' values

Respond with a JSON object in this exact format:
{
  \"translations\": [
    {
      \"key\": \"original_key_1\",
      \"translation\": \"translated_text_1\"
    },
    {
      \"key\": \"original_key_2\", 
      \"translation\": \"translated_text_2\"
    }
  ]
}

Important:
- Respond only with valid JSON
- Do not include explanations or additional fields";
    }
    
    private function buildBatchProofreadingPrompt($items, $language) {
        $itemsJson = json_encode($items, JSON_PRETTY_PRINT);
        
        return "Review the following localized texts for a mobile/desktop application. Evaluate the quality and provide feedback for each.

Input data structure:
{$itemsJson}

Language: {$language}

For each text, consider:
- Grammar and spelling
- Clarity and naturalness  
- Appropriateness for UI context
- Consistency with typical app terminology
- Cultural appropriateness

Respond with a JSON object in this exact format:
{
  \"reviews\": [
    {
      \"key\": \"original_key_1\",
      \"status\": \"good\",
      \"feedback\": \"explanation or empty string if good\"
    },
    {
      \"key\": \"original_key_2\",
      \"status\": \"wording\", 
      \"feedback\": \"explanation of suggestions\"
    }
  ]
}

Status meanings:
- \"good\": The text is well-written and appropriate
- \"wording\": The text is understandable but could be improved with better wording
- \"issue\": There are serious problems (grammar, unclear meaning, inappropriate tone, etc.)

Important: Respond only with valid JSON, no other text.";
    }
}
?>

File: backend/src/main/java/com/xcstring/editor/service/AIService.java
package com.xcstring.editor.service;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.xcstring.editor.config.AppProperties;
import com.xcstring.editor.config.AppProperties.ProviderAiConfig;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;

@Service
@Slf4j
public class AIService {

    private final AppProperties appProperties;
    private final PromptService promptService;
    private final Gson gson = new GsonBuilder().create();

    public AIService(AppProperties appProperties, PromptService promptService) {
        this.appProperties = appProperties;
        this.promptService = promptService;
    }

    public boolean isEnabled() {
        return appProperties.getAi().isEnabled();
    }

    private boolean isDeepL(String provider) {
        return "deepl".equals(provider);
    }

    public Map<String, Object> getAvailableProviders() {
        Map<String, Object> providers = new LinkedHashMap<>();
        
        if (!isEnabled()) {
            return providers;
        }

        AppProperties.AiProps aiProps = appProperties.getAi();
        
        addProviderIfAvailable(providers, "openai", aiProps.getOpenai());
        addProviderIfAvailable(providers, "anthropic", aiProps.getAnthropic());
        addProviderIfAvailable(providers, "zai", aiProps.getZai());
        addProviderIfAvailable(providers, "deepl", aiProps.getDeepl());
        addProviderIfAvailable(providers, "openai_compatible", aiProps.getOpenaiCompatible());

        return providers;
    }

    private void addProviderIfAvailable(Map<String, Object> providers, String name, ProviderAiConfig config) {
        if (config.isEnabled() && config.getApiKey() != null && !config.getApiKey().isEmpty()) {
            Map<String, Object> providerInfo = new LinkedHashMap<>();
            providerInfo.put("name", name);
            providerInfo.put("models", config.getModels());
            providerInfo.put("capabilities", isDeepL(name)
                ? Collections.singletonList("translate")
                : Arrays.asList("translate", "proofread"));
            providers.put(name, providerInfo);
        }
    }

    private static class ResolvedProvider {
        final String provider;
        final String model;
        final ProviderAiConfig config;

        ResolvedProvider(String provider, String model, ProviderAiConfig config) {
            this.provider = provider;
            this.model = model;
            this.config = config;
        }
    }

    private ResolvedProvider resolveProvider(String provider, String model) {
        AppProperties.AiProps aiProps = appProperties.getAi();
        
        if (provider == null) {
            provider = aiProps.getDefaultProvider();
        }
        if (model == null) {
            model = aiProps.getDefaultModel();
        }

        ProviderAiConfig config = getProviderConfig(provider);
        
        if (config == null || !config.isEnabled()) {
            throw new RuntimeException("Provider " + provider + " is not available");
        }

        return new ResolvedProvider(provider, model, config);
    }

    private ProviderAiConfig getProviderConfig(String provider) {
        AppProperties.AiProps aiProps = appProperties.getAi();
        
        switch (provider) {
            case "openai":
                return aiProps.getOpenai();
            case "anthropic":
                return aiProps.getAnthropic();
            case "zai":
                return aiProps.getZai();
            case "deepl":
                return aiProps.getDeepl();
            case "openai_compatible":
                return aiProps.getOpenaiCompatible();
            default:
                return null;
        }
    }

    private Object makeAIRequest(String provider, String model, String prompt,
                                  ProviderAiConfig config, boolean expectJson) {
        switch (provider) {
            case "openai":
            case "openai_compatible":
                return makeOpenAIRequest(model, prompt, config, expectJson);
            case "anthropic":
                return makeAnthropicRequest(model, prompt, config, expectJson);
            case "zai":
                return makeZAIRequest(model, prompt, config, expectJson);
            default:
                throw new RuntimeException("Unsupported provider: " + provider);
        }
    }

    private OkHttpClient buildClient(int timeoutSeconds) {
        return new OkHttpClient.Builder()
            .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .build();
    }

    @SuppressWarnings("unchecked")
    private Object makeOpenAIRequest(String model, String prompt, ProviderAiConfig config, boolean expectJson) {
        int timeout = expectJson ? 120 : 30;
        OkHttpClient client = buildClient(timeout);

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("model", model);
        data.put("input", prompt);
        data.put("max_tokens", expectJson ? 8000 : 500);
        data.put("temperature", 0.3);

        if (expectJson) {
            Map<String, String> responseFormat = new LinkedHashMap<>();
            responseFormat.put("type", "json_object");
            data.put("response_format", responseFormat);
        }

        String url = config.getBaseUrl() + "/responses";

        try {
            String jsonBody = gson.toJson(data);
            
            Request request = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("Authorization", "Bearer " + config.getApiKey())
                .post(RequestBody.create(jsonBody, MediaType.parse("application/json")))
                .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new RuntimeException("AI request failed: HTTP " + response.code());
                }

                String responseBody = response.body().string();
                Map<String, Object> result = gson.fromJson(responseBody, new TypeToken<Map<String, Object>>(){}.getType());
                
                if (result == null || result.get("output") == null) {
                    throw new RuntimeException("Invalid AI response format");
                }

                List<Map<String, Object>> output = (List<Map<String, Object>>) result.get("output");
                if (output.isEmpty() || output.get(0).get("content") == null) {
                    throw new RuntimeException("Invalid AI response format");
                }

                List<Map<String, Object>> content = (List<Map<String, Object>>) output.get(0).get("content");
                if (content.isEmpty() || content.get(0).get("text") == null) {
                    throw new RuntimeException("Invalid AI response format");
                }

                String contentText = (String) content.get(0).get("text");

                if (expectJson) {
                    return gson.fromJson(contentText, new TypeToken<Map<String, Object>>(){}.getType());
                }

                return contentText;
            }
        } catch (IOException e) {
            throw new RuntimeException("AI request failed: " + e.getMessage(), e);
        }
    }

    @SuppressWarnings("unchecked")
    private Object makeAnthropicRequest(String model, String prompt, ProviderAiConfig config, boolean expectJson) {
        int timeout = expectJson ? 120 : 30;
        OkHttpClient client = buildClient(timeout);

        String actualPrompt = prompt;
        if (expectJson) {
            actualPrompt = prompt + "\n\nIMPORTANT: Respond only with valid JSON, no other text.";
        }

        Map<String, Object> message = new LinkedHashMap<>();
        message.put("role", "user");
        message.put("content", actualPrompt);

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("model", model);
        data.put("max_tokens", expectJson ? 40000 : 500);
        data.put("temperature", 0.3);
        data.put("messages", Collections.singletonList(message));

        String url = config.getBaseUrl() + "/v1/messages";

        try {
            String jsonBody = gson.toJson(data);
            
            Request request = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("x-api-key", config.getApiKey())
                .addHeader("anthropic-version", "2023-06-01")
                .post(RequestBody.create(jsonBody, MediaType.parse("application/json")))
                .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new RuntimeException("AI request failed: HTTP " + response.code());
                }

                String responseBody = response.body().string();
                Map<String, Object> result = gson.fromJson(responseBody, new TypeToken<Map<String, Object>>(){}.getType());
                
                if (result == null || result.get("content") == null) {
                    throw new RuntimeException("Invalid AI response format");
                }

                List<Map<String, Object>> content = (List<Map<String, Object>>) result.get("content");
                if (content.isEmpty() || content.get(0).get("text") == null) {
                    throw new RuntimeException("Invalid AI response format");
                }

                String contentText = ((String) content.get(0).get("text")).trim();

                if (expectJson) {
                    return gson.fromJson(contentText, new TypeToken<Map<String, Object>>(){}.getType());
                }

                return contentText;
            }
        } catch (IOException e) {
            throw new RuntimeException("AI request failed: " + e.getMessage(), e);
        }
    }

    @SuppressWarnings("unchecked")
    private Object makeZAIRequest(String model, String prompt, ProviderAiConfig config, boolean expectJson) {
        int timeout = expectJson ? 120 : 30;
        OkHttpClient client = buildClient(timeout);

        Map<String, Object> message = new LinkedHashMap<>();
        message.put("role", "user");
        message.put("content", prompt);

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("model", model);
        data.put("messages", Collections.singletonList(message));
        data.put("max_tokens", expectJson ? 40000 : 500);
        data.put("temperature", 0.3);

        if (expectJson) {
            Map<String, String> responseFormat = new LinkedHashMap<>();
            responseFormat.put("type", "json_object");
            data.put("response_format", responseFormat);
        }

        String url = config.getBaseUrl() + "/chat/completions";

        try {
            String jsonBody = gson.toJson(data);
            
            Request request = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("Authorization", "Bearer " + config.getApiKey())
                .post(RequestBody.create(jsonBody, MediaType.parse("application/json")))
                .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new RuntimeException("Z.AI request failed: HTTP " + response.code());
                }

                String responseBody = response.body().string();
                Map<String, Object> result = gson.fromJson(responseBody, new TypeToken<Map<String, Object>>(){}.getType());
                
                if (result == null || result.get("choices") == null) {
                    throw new RuntimeException("Invalid Z.AI response format");
                }

                List<Map<String, Object>> choices = (List<Map<String, Object>>) result.get("choices");
                if (choices.isEmpty() || choices.get(0).get("message") == null) {
                    throw new RuntimeException("Invalid Z.AI response format");
                }

                Map<String, Object> messageResult = (Map<String, Object>) choices.get(0).get("message");
                if (messageResult.get("content") == null) {
                    throw new RuntimeException("Invalid Z.AI response format");
                }

                String contentText = ((String) messageResult.get("content")).trim();

                if (expectJson) {
                    return gson.fromJson(contentText, new TypeToken<Map<String, Object>>(){}.getType());
                }

                return contentText;
            }
        } catch (IOException e) {
            throw new RuntimeException("Z.AI request failed: " + e.getMessage(), e);
        }
    }

    private String getDeepLBaseUrl(ProviderAiConfig config) {
        if (config.getBaseUrl() != null && !config.getBaseUrl().isEmpty()) {
            return config.getBaseUrl();
        }
        
        String apiKey = config.getApiKey();
        if (apiKey != null && apiKey.endsWith(":fx")) {
            return "https://api-free.deepl.com";
        }
        return "https://api.deepl.com";
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> callDeepLApi(List<String> texts, String sourceLanguage, 
                                                    String targetLanguage, String model, 
                                                    ProviderAiConfig config) {
        int timeout = texts.size() > 1 ? 120 : 30;
        OkHttpClient client = buildClient(timeout);
        
        String baseUrl = getDeepLBaseUrl(config);

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("text", texts);
        data.put("target_lang", targetLanguage.toUpperCase());
        data.put("model_type", model);
        data.put("formality", "prefer_less");
        data.put("split_sentences", "0");

        if (sourceLanguage != null && !sourceLanguage.isEmpty()) {
            data.put("source_lang", sourceLanguage.toUpperCase());
        }

        String url = baseUrl + "/v2/translate";

        try {
            String jsonBody = gson.toJson(data);
            
            Request request = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("Authorization", "DeepL-Auth-Key " + config.getApiKey())
                .post(RequestBody.create(jsonBody, MediaType.parse("application/json")))
                .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new RuntimeException("DeepL API request failed: HTTP " + response.code());
                }

                String responseBody = response.body().string();
                Map<String, Object> result = gson.fromJson(responseBody, new TypeToken<Map<String, Object>>(){}.getType());
                
                if (result == null || result.get("translations") == null) {
                    throw new RuntimeException("Invalid DeepL API response format");
                }

                return (List<Map<String, Object>>) result.get("translations");
            }
        } catch (IOException e) {
            throw new RuntimeException("DeepL API request failed: " + e.getMessage(), e);
        }
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> deepLBatchTranslate(List<Map<String, Object>> items, 
                                                          String sourceLanguage, 
                                                          String targetLanguage, 
                                                          String model, 
                                                          ProviderAiConfig config) {
        List<String> texts = new ArrayList<>();
        for (Map<String, Object> item : items) {
            texts.add((String) item.get("text"));
        }

        List<Map<String, Object>> translations = callDeepLApi(texts, sourceLanguage, 
            targetLanguage, model, config);

        List<Map<String, Object>> result = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            Map<String, Object> resultMap = new LinkedHashMap<>();
            resultMap.put("key", items.get(i).get("key"));
            
            String translation = "";
            if (i < translations.size() && translations.get(i).get("text") != null) {
                translation = ((String) translations.get(i).get("text")).trim();
            }
            resultMap.put("translation", translation);
            
            result.add(resultMap);
        }

        return result;
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> batchTranslate(List<Map<String, Object>> items, 
                                                    String sourceLanguage, 
                                                    String targetLanguage, 
                                                    String provider, 
                                                    String model) {
        if (!isEnabled()) {
            throw new RuntimeException("AI features are not enabled");
        }

        if (items == null || items.isEmpty()) {
            return Collections.emptyList();
        }

        ResolvedProvider resolved = resolveProvider(provider, model);

        if (isDeepL(resolved.provider)) {
            return deepLBatchTranslate(items, sourceLanguage, targetLanguage, 
                resolved.model, resolved.config);
        }

        String prompt = buildBatchTranslationPrompt(items, sourceLanguage, targetLanguage);
        Object result = makeAIRequest(resolved.provider, resolved.model, prompt, 
            resolved.config, true);

        Map<String, Object> resultMap = (Map<String, Object>) result;
        if (resultMap.get("translations") == null) {
            throw new RuntimeException("Invalid batch translation response format");
        }

        return (List<Map<String, Object>>) resultMap.get("translations");
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> batchProofread(List<Map<String, Object>> items, 
                                                    String language, 
                                                    String provider, 
                                                    String model) {
        if (!isEnabled()) {
            throw new RuntimeException("AI features are not enabled");
        }

        if (items == null || items.isEmpty()) {
            return Collections.emptyList();
        }

        ResolvedProvider resolved = resolveProvider(provider, model);

        if (isDeepL(resolved.provider)) {
            throw new RuntimeException("DeepL does not support proofreading. Please select a different provider.");
        }

        String prompt = buildBatchProofreadingPrompt(items, language);
        Object result = makeAIRequest(resolved.provider, resolved.model, prompt, 
            resolved.config, true);

        Map<String, Object> resultMap = (Map<String, Object>) result;
        if (resultMap.get("reviews") == null) {
            throw new RuntimeException("Invalid batch proofreading response format");
        }

        return (List<Map<String, Object>>) resultMap.get("reviews");
    }

    private String buildBatchTranslationPrompt(List<Map<String, Object>> items, 
                                                String sourceLanguage, 
                                                String targetLanguage) {
        String itemsJson;
        try {
            itemsJson = gson.toJson(items);
        } catch (Exception e) {
            itemsJson = items.toString();
        }

        return promptService.renderTranslationPrompt(sourceLanguage, targetLanguage, itemsJson);
    }

    private String buildBatchProofreadingPrompt(List<Map<String, Object>> items, String language) {
        String itemsJson;
        try {
            itemsJson = gson.toJson(items);
        } catch (Exception e) {
            itemsJson = items.toString();
        }

        return promptService.renderProofreadingPrompt(language, itemsJson);
    }
}
